Fix issue with second derivative values on faces.

jwpeterson opened the pull request:
This PR makes it possible to get test and trial function second derivative values from `MooseVariableInterface` and fixes an issue where second derivatives were not being properly zeroed on faces in Steady simulations.  Also adds a test to ensure that this functionality keeps working.

Thanks to @jiangwen84 and @andrsd for their assistance and example code.

Closes #6327.

jwpeterson committed:
MooseVariableInterface: add secondTestFace() and secondPhiFace().

These need to be called in e.g. boundary conditions instead of secondTest() and secondPhi().

Refs #6327.

jwpeterson committed:
Fix incorrect zeroing of _second_u in MooseVariable.

Refs #6327.

jwpeterson committed:
Add test of second derivatives in IntegratedBC.

Code originally by: Wen Jiang <wen.jiang@inl.gov>

Refs #6327.

dschwen commented on framework/src/base/MooseVariableInterface.C:
```diff
@@ -155,6 +155,16 @@ MooseVariableInterface::secondTest()
+  if (_nodal)
+    mooseError("Nodal variables do not have second derivatives");
+
+  return const_cast<VariableTestSecond &>(_variable->secondPhiFace());
```
Why don't you make that method return a const reference? We should switch all of these methods over at some time.

jwpeterson replied:
I just assumed we were doing this for some reason.  I think I'd rather update them all in a separate PR anyway -- it's going to change a lot of user code when we do...

dschwen replied:
Yeah, I guess having a mix of const and non-const would be weird.

jwpeterson commented:
Hm... I guess I can see why this might change the results of Laplacian jump indicator since it uses `_second_u_neighbor`, but not sure why that would only be uncovered by the threaded runs...

jwpeterson commented:
Nevermind, nothing to do with threads... that's just the name of the target.  I think the `LaplacianJumpIndicator` just needs to be regolded.

jwpeterson commented:
Yeah, so the error indicator really made no sense before: the exact solution to this test problem is nearly linear, so the Laplacian error jump should be very small for Hermite elements.  Instead the error looked like this:

![before](https://cloud.githubusercontent.com/assets/1775907/13152295/afcb1b6c-d62b-11e5-8568-32d184dd4b5d.png)


But now looks like this (note the scale):

![after](https://cloud.githubusercontent.com/assets/1775907/13152336/d835013a-d62b-11e5-9e4e-4c71b5ad3a91.png)

jwpeterson committed:
Regold the LaplacianJumpIndicator test.

After fixing the issue with _second_u_neighbor not being properly
zeroed, it changed the results of the error indicator in this test
(for the better).

I also tightened down the nl_rel_tol so we will hopefully get the same
results in parallel and serial.

Refs #6327.

jwpeterson commented:
Damn, really small difference in the error indicator as computed in serial vs. parallel now:
```
indicators/laplacian_jump_indicator.test: Element variables:
indicators/laplacian_jump_indicator.test:    error  rel diff:  6.4447937e-06 ~  6.4547313e-06 = 1.53958e-03 (block 0, elmt 45)
```

permcody commented:
:+1:
